### PR TITLE
KEYCLOAK-11733 Add endpoint to get parents of a role

### PR DESCRIPTION
**Pull request by amdev74**
https://issues.jboss.org/browse/KEYCLOAK-11733

### CODE REVIEW

File: model/infinispan/src/main/java/org/keycloak/models/cache/infinispan/RoleAdapter.java [hunk 1]
```diff
@@ -141,6 +142,30 @@ public void removeCompositeRole(RoleModel role) {
...
+    
+	@Override
+	public void addParentRole(RoleModel role) {
+        getDelegateForUpdate();
+        updated.addParentRole(role);
+	}
+    
+	@Override
+	public Set<RoleModel> getParents() {
+        if (isUpdated()) return updated.getParents();
```
Comment (Cracky5457) [inline model/infinispan/src/main/java/org/keycloak/models/cache/infinispan/RoleAdapter.java]: Issue with the cache, still returning roles after unassign them

Comment (Cracky5457) [reply model/infinispan/src/main/java/org/keycloak/models/cache/infinispan/RoleAdapter.java]: fixed

**Comment by stianst** [comment]
I'm afraid we'll have to put this on hold due to other priorities and the potential issues it could cause to include. So we'll have to get back to this one in a few months.

**Comment by Cracky5457** [comment]
Ok no problem, I don't. think the is a lot of risk with this one, it's just an unidirectional JPA mapping which is now bidirectional. I'm using it on our custom release since month with no issue.

**Comment by stianst** [comment]
Looking at this again I'm not all that in favour of adding it. At least not in its current form.

There's no parent/child relationship for roles as such this is more of a search to find composite roles that contain the role rather than getting "parents" of the role.

Can you elaborate on the use-case why this is important?

**Comment by stianst** [comment]
oh, and sorry for the absolutely ridiculous time it's taken to comment on this...

**Comment by Cracky5457** [comment]
I agree that `parents` is maybe not the right terminology we can maybe found an other one.

I can give you our use-case

We have groups of people managed by role in a `group` client for exemple ( we did the choice of not using the real KeycloakGroup structure to do that for severals reasons but we are using them for an other use case )

so let sey I have `groupA` with cracky5457 inside ( so cracky5457 have the role groupA )

Also we have `groupB` with stianst inside ( so stianst have the role groupB )
Let say we have also `groupC` with other people inside.

So a user can be member or administrator of a group ( we have admin roles too ), but a group can also be member of an other group !

So let say `groupB` and `groupC` can be member of `groupA`, so ``groupC` will be composite with `groupA` and `groupB` will also be composite with `groupA` meaning that if you have `groupB` you also avec the role `groupA`, so stianst for exemple you are directly member of groupB et now indirectly member of `groupA`

And so we have a UI to visualize who is member and which group are member of groupA, we need to see who have the role `groupA` so we display the member and the group, then we can navigate to the tree to see who is member of groupB ( can be some users and some other groups etc. ). 

So to have this informations is this new endpoint`groupA/parents` which will return `groupB` and `groupC`

for KeycloakUsers and KeycloakGroup we have a kind of equivalents

`/clients/{clientId}/roles/{roleName}/users` return KeycloakUser for a given Role
`/clients/{clientId}/roles/{roleName}/groups` return KeycloakGroup for a given Role

the new endpoint
`/clients/{clientId}/roles/{roleName}/parents` return RoleRepresentation

( they also works with `{roleById}/users|groups|parents )

At the end it's a pretty powerfull and flexible groups managing system and it's the kind of thing you can do with this endpoint.

Also I have an other but very complicated merge request here https://github.com/keycloak/keycloak/pull/6326

Which also need the parent api to perform a deep search to fetch which users have a role even from an undirect assignation, so it's an other use-case.

It's probably a pretty advanced usage of client roles. No problem if you are not interested we can keep it in our fork.

**Comment by stianst** [comment]
Could you send an email to the dev mailing list around this? I don't have any strong objections of adding this, but would definitively want the language to be changed as there is no parent/child relationship with composite roles, as it's a many-to-many relationship. Not quite sure about the way it's implemented here, but haven't had time to dig into it properly.

**Comment by Cracky5457** [comment]
In the first time I used `parents` because the composite roles was already named `children` in the code which is also pretty wrong but lets keep children.

We can see the roles in keycloak like some kind of nested role, it could be represent by a `graph`.

In the `graph theory` the correct word would probably be `predecessor` role or some kind of `ascending` role

**Comment by stianst** [comment]
Not sure what I prefer to be honest. I sent an email to the keycloak-dev mailing list to see what others think.

**Comment by stianst** [comment]
@Cracky5457 sorry for the long time it's taken to get this reviewed and progressed. I was looking a bit more at this today as we are not trying to get through the backlog of PRs.

I'm okay with using the name 'parents' as I didn't get any feedback on dev ML.

One concern I have here though is that this effectively results in loading parent roles for all use of roles. For any other use-case than fetching parent roles from admin API we don't need to fetch the parent roles, for example during a login. This will add overhead with no benefits in 99.9% cases. A simpler and perhaps better idea might be to just not cache the parent roles at all and just allow fetching/querying them from admin APIs.

**Comment by stianst** [comment]
@Cracky5457 I'm going to close this for now, if you are still interested in this feel free to reopen it. To add this we'd need to consider how it can be done without performance penalty to regular requests.

**Pull request closed without merging**